core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -283,6 +283,44 @@ protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotI
     }
   }
 
+  /**
+   * Validates that no delete files matching a filter have been added to the table since a starting snapshot.
+   *
+   * @param base table metadata to validate
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param conflictDetectionFilter an expression used to find new conflicting delete files
+   * @param caseSensitive whether expression evaluation should be case sensitive
+   */
+  protected void validateAddedDeleteFiles(TableMetadata base, Long startingSnapshotId,
+                                          Expression conflictDetectionFilter, boolean caseSensitive) {
+    // if there is no current table state, no files have been added
+    if (base.currentSnapshot() == null) {
+      return;
+    }
+
+    Pair<List<ManifestFile>, Set<Long>> history =
+        validationHistory(base, startingSnapshotId, VALIDATE_ADDED_DELETE_FILES_OPERATIONS, ManifestContent.DELETES);
+    List<ManifestFile> deleteManifests = history.first();
+
+    DeleteFileIndex.Builder deleteIndexBuilder = DeleteFileIndex.builderFor(ops.io(), deleteManifests)
+        .filterData(conflictDetectionFilter)
```

Comment:
This is a good optimization. I think we should see if we can add it to the file-based check as well.

Reply:
Looks like that's https://github.com/apache/iceberg/pull/3071. Thanks for fixing it!

Reply:
Done and merged. Resolving this.